datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -415,6 +415,18 @@ impl PartialEq<dyn Any> for InListExpr {
     }
 }
 
+/// Checks if two types are logically equal, dictionary types are compared by their value types.
+fn is_logically_eq(lhs: &DataType, rhs: &DataType) -> bool {
```

Comment:
hm, I think we have a similar method in https://github.com/apache/arrow-datafusion/blob/f55c1d90215614ce531a4103c7dbebf318de1cfd/datafusion/common/src/dfschema.rs#L643 basically the schema struct has a lot of equal checks between datatypes. may want to move them into separate utils struct as they mostly working with datatypes rather than schema

Reply:
I agree we could use `datatype_is_logically_equal` instead here. It also seems like a good idea to make that function more discoverable -- I also think it would be fine to do as a follow on PR